pimd/pim_autorp.c

```diff
@@ -0,0 +1,1146 @@
+		/* If hold time is zero, make sure there doesn't exist a hold timer for it already */
+		event_cancel(&trp->hold_timer);
+	}
+
+	return 0;
+}
+
+static int pim_autorp_discovery(struct pim_autorp *autorp, uint8_t rpcnt,
+				uint16_t holdtime, char *buf, size_t buf_size)
+{
+	int i, j;
+	struct autorp_pkt_rp *rp;
+	struct autorp_pkt_grp *grp;
+	size_t offset = 0;
+	pim_addr rp_addr;
+	struct prefix grppfix;
+	char plname[32];
+	struct prefix_list *pl;
+	struct prefix_list_entry *ple;
+	int64_t seq = 1;
+	int ret = 0;
+
+	for (i = 0; i < rpcnt; ++i) {
+		if ((buf_size - offset) < AUTORP_RPLEN)
+			return -1;
+
+		rp = (struct autorp_pkt_rp *)(buf + offset);
+		offset += AUTORP_RPLEN;
+
+		rp_addr.s_addr = rp->addr;
+		/* TODO: What so do with pim version?
+		 * rp->pimver;
+		 */
+
+		if (rp->grpcnt == 0) {
+			/* No groups?? */
+			zlog_warn("%s: Discovery message has no groups for RP %pI4",
+				  __func__, (in_addr_t *)&(rp->addr));
+			continue;
+		}
+
+		if ((buf_size - offset) < AUTORP_GRPLEN)
+			return -1;
+
+		grp = (struct autorp_pkt_grp *)(buf + offset);
+		offset += AUTORP_GRPLEN;
+
+		if (rp->grpcnt == 1 && grp->negprefix == 0) {
+			/* Only one group with positive prefix, we can use the standard RP API */
+			grppfix.family = AF_INET;
+			grppfix.prefixlen = grp->masklen;
+			grppfix.u.prefix4.s_addr = grp->addr;
+			ret += pim_autorp_add_rp(autorp, rp_addr, grppfix, NULL,
+						 holdtime);
+		} else {
+			/* More than one grp, or the only group is a negative prefix, need to make a prefix list for this RP */
+			snprintfrr(plname, sizeof(plname), "__AUTORP_%pI4__",
+				   &rp_addr);
+			pl = prefix_list_get(AFI_IP, 0, plname);
+
+			for (j = 0; j < rp->grpcnt; ++j) {
+				/* grp is already pointing at the first group in the buffer */
+				ple = prefix_list_entry_new();
+				ple->pl = pl;
+				ple->seq = seq;
+				seq += 5;
+				memset(&ple->prefix, 0, sizeof(ple->prefix));
+				prefix_list_entry_update_start(ple);
+				ple->type = (grp->negprefix ? PREFIX_DENY
+							    : PREFIX_PERMIT);
+				ple->prefix.family = AF_INET;
+				ple->prefix.prefixlen = grp->masklen;
+				ple->prefix.u.prefix4.s_addr = grp->addr;
+				ple->any = false;
+				ple->ge = 0;
+				ple->le = 32;
+				prefix_list_entry_update_finish(ple);
+
+				if ((buf_size - offset) < AUTORP_GRPLEN)
+					return -1;
+
+				grp = (struct autorp_pkt_grp *)(buf + offset);
+				offset += AUTORP_GRPLEN;
+			}
+
+			ret += pim_autorp_add_rp(autorp, rp_addr, grppfix,
+						 plname, holdtime);
+		}
+	}
+
+	if (PIM_DEBUG_AUTORP)
+		zlog_debug("%s: Processed AutoRP Discovery message", __func__);
+
+	return ret;
+}
+
+static int pim_autorp_msg(struct pim_autorp *autorp, char *buf, size_t buf_size)
+{
+	struct autorp_pkt_hdr *h;
+
+	if (buf_size < AUTORP_HDRLEN)
+		return -1;
+
+	h = (struct autorp_pkt_hdr *)buf;
+
+	if (h->version != AUTORP_VERSION)
+		return -1;
+
+	if (h->type == AUTORP_ANNOUNCEMENT_TYPE &&
+	    pim_autorp_announcement(autorp, h->rpcnt, htons(h->holdtime),
```

Comment:
The reason I am asking for a bool here is that a -1 as a failure without the function name giving us a clue implies that success is a failure.